qiskit/circuit/quantumcircuit.py

```diff
@@ -2209,18 +2210,31 @@ def remove_final_measurements(self, inplace: bool = True) -> Optional["QuantumCi
         remove_final_meas = RemoveFinalMeasurements()
         new_dag = remove_final_meas.run(dag)
 
-        # Set circ cregs and instructions to match the new DAGCircuit's
+        # Filter only cregs/clbits still in new DAG, preserving original circuit order
+        circ.cregs = [creg for creg in circ.cregs if creg in new_dag.cregs.values()]
+        circ._clbits = [clbit for clbit in circ._clbits if clbit in new_dag.clbits]
+
+        creg_bits = {clbit for creg in circ.cregs for clbit in creg}
+
+        # Recalculate clbit indicies
+        circ._clbit_indices = {
+            clbit: BitLocations(
+                i, [reg for reg in circ._clbit_indices[clbit].registers if reg[0] in circ.cregs]
+            )
+            for i, clbit in enumerate(circ.clbits)
+            if clbit in creg_bits
+        }
```

Comment:
Ah, this is the comment I thought I'd lost.

Reply:
Ah yeah, duplicating the bit indices calculation here seemed like the cleanest way to preserve the original ordering for both `cregs` and `clbits` at the time (originally I in 64f137394c353bc74417eee8b8baeae32e20ea83 I was not preserving order). Though thinking about it a little more now, I believe I can keep the correct ordering while also making use of the `add_(bits|cregs)` functions. I will make this change and come back to this comment with its commit ID.

Reply:
Done in a1ff147038c289483f157df75ba42c869d96f5b7. Turns out we can easily keep the order by just adding all the `clbits` first, and then adding in the `cregs`.

note: in a later commit, I replaced the `clear()` calls with assignments of new collections, as you have in your example. I'm assuming this is to avoid clearing them out from under users who may have a reference.